addons/sale/models/sale_order.py

```diff
@@ -1004,9 +1066,23 @@ def _should_be_locked(self):
         user = self[:1].create_uid
         return user and user.sudo().has_group('sale.group_auto_done_setting')
 
-    def _can_be_confirmed(self):
-        self.ensure_one()
-        return self.state in {'draft', 'sent'}
+    def _check_can_be_confirmed(self):
+        """ Check whether order can be confirmed or not if not then raise error. """
+        for order in self:
+            if order.state not in {'draft', 'sent'}:
+                raise UserError(_(
+                    "The following orders are not in a state requiring confirmation: %s",
+                    ", ".join(self.mapped('display_name')),
+                ))
+            if any(
+                not line.display_type
+                and not line.is_downpayment
+                and not line.product_id
+                for line in order.order_line
+            ):
+                raise ValidationError(_(
+                    "A line on these orders missing a product, you cannot confirm it."
+                ))
```

Comment:
Isn't this already enforced by the SQL constraint on `sale.order.line` ?

Reply:
No it isn't, There is not constraint like that on SOL :thinking: 
Adding one would prevent extraction :see_no_evil:

Reply:
```python
        ('accountable_required_fields',
            "CHECK(display_type IS NOT NULL OR is_downpayment OR (product_id IS NOT NULL AND product_uom IS NOT NULL))",
            "Missing required fields on accountable sale order line."),
```
:thinking: 

Also, your changes here are probably the cause of the failing query count in `sale_subcription` tests, you have to update the query count (or find a way to avoid those additional queries when they're not necessary).

Reply:
I am not sure why that constraint is not triggering :slightly_frowning_face: 

Updating query count in `sale_subscription` (We really don't want to add this error here but it was allowing SO to confirm even some SOL does not have product field set on them also UX require attribute was also not triggering because sol is not in edition mode :cry: )

Reply:
Yeah, it's strange it's not working :confused:
Let's update of the query count for now.
Could you update the method so that it returns a boolean `_can_be_confirmed()`, and only raise the ValidationError in `action_confirm` if the returned value is falsy?  That way, other modules can override the check more easily ;).

Reply:
I did this to have different error message depending on issue
Is it okay to have same error message :thinking: 
(

Reply:
Could work :ok_hand: (if the strings are correctly translated :eyes:).